bftools: init at 5.9.2

tbenst opened the pull request:
###### Motivation for this change
This package is useful for microscopy. First pull request here so feedback most welcome!

###### Things done


- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I tested bfconvert, which is the only tool I use, and this worked properly for converting a .oir file to a .ome.btf. The other executables display "--help" properly but were not otherwise tested.

---

tbenst committed:
bftools: init at 5.9.2

veprbl commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  # phases = [ "unpackPhase" "installPhase" "fixupPhase" ];
+
+  meta = with stdenv.lib; {
```
Need to set `meta.platforms`, that's important for Hydra

tbenst replied:
Should this be `platforms = platforms.linux ++ platforms.darwin;` or `platforms = platforms.all;` or `platforms = platforms.unix;` or something else? Presumably should be pretty portable thanks to java but have only tested NixOS and Ubuntu

veprbl replied:
`platforms.all` is good. Your expression will only evaluate where `openjdk` evaluates.

veprbl commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  buildInputs = [ unzip pkgs.makeWrapper ];
+
+  # phases = [ "unpackPhase" "installPhase" "fixupPhase" ];
```
Remove this

veprbl commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,36 @@
+        for script in "$out"/bin/*; do
+          wrapProgram "$script" --prefix PATH : "${lib.makeBinPath [ openjdk ]}"
+        done
+      '';
```
Fix indentation here, please

symphorien commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,36 @@
+        done
+      '';
+
+  buildInputs = [ unzip pkgs.makeWrapper ];
```
replace pkgs with makeWrapper in the arguments in the first line of the file and then you can refer to makeWrapper without `pkgs.`.

tbenst replied:
done thx

symphorien commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,36 @@
+  # phases = [ "unpackPhase" "installPhase" "fixupPhase" ];
+
+  meta = with stdenv.lib; {
+    description = "A bundle of scripts for using Bio-Formats on the command line with bioformats_package.jar already included.";
```
This should not end with a period according to https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md

veprbl commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  postFixup = ''
+        for script in "$out"/bin/*; do
+          wrapProgram "$script" --prefix PATH : "${lib.makeBinPath [ openjdk ]}"
```
You only need to wrap bf.sh

tbenst replied:
good catch thx!

tbenst replied:
I'm sure I'm missing something basic, but I get this error: `Builder called die: Cannot wrap '/nix/store/p6mycdqcysgpkmi5rhd797ip2fssbppq-bftools-5.9.2/bin/bf.sh' because it is not an executable file`. Should I use something other than wrapProgram?

veprbl replied:
Add the executable permissions then?

symphorien commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp ./* $out/
```
Is there are reason to put things at the root of the derivation ? `$out/bin$` should be enough.

tbenst replied:
There are a number of other .sh files, and one .jar file, and the executable expect to be in the same directory. Didn't think that all these files should go in bin? Let me know

Here are the unzipped files:
```
-rw-r--r--  1 tyler tyler 2.1K Jul  3  2018 bf.bat
-rwxr-xr-x  1 tyler tyler  374 Jul  3  2018 bfconvert
-rw-r--r--  1 tyler tyler  298 Jul  3  2018 bfconvert.bat
-rwxr-xr-x  1 tyler tyler 2.1K Jul  3  2018 bf.sh
-rw-r--r--  1 tyler tyler  31M Sep  3 10:51 bioformats_package.jar
-rw-r--r--  1 tyler tyler 1.2K Jul  3  2018 config.bat
-rw-r--r--  1 tyler tyler 1.1K Jul  3  2018 config.sh
-rwxr-xr-x  1 tyler tyler  375 Jul  3  2018 domainlist
-rw-r--r--  1 tyler tyler  299 Jul  3  2018 domainlist.bat
-rwxr-xr-x  1 tyler tyler  379 Jul  3  2018 formatlist
-rw-r--r--  1 tyler tyler  303 Jul  3  2018 formatlist.bat
-rwxr-xr-x  1 tyler tyler 1.5K Jul  3  2018 ijview
-rw-r--r--  1 tyler tyler 1.3K Jul  3  2018 ijview.bat
-rw-r--r--  1 tyler tyler  12K Jan 10 11:59 .ijview.swp
-rw-r--r--  1 tyler tyler  281 Jul  3  2018 logback.xml
-rwxr-xr-x  1 tyler tyler  401 Jul  3  2018 mkfake
-rw-r--r--  1 tyler tyler  319 Jul  3  2018 mkfake.bat
-rwxr-xr-x  1 tyler tyler  428 Jul  3  2018 showinf
-rw-r--r--  1 tyler tyler  358 Jul  3  2018 showinf.bat
-rwxr-xr-x  1 tyler tyler  418 Jul  3  2018 tiffcomment
-rw-r--r--  1 tyler tyler  348 Jul  3  2018 tiffcomment.bat
-rwxr-xr-x  1 tyler tyler  356 Jul  3  2018 xmlindent
-rw-r--r--  1 tyler tyler  280 Jul  3  2018 xmlindent.bat
-rwxr-xr-x  1 tyler tyler  352 Jul  3  2018 xmlvalid
-rw-r--r--  1 tyler tyler  276 Jul  3  2018 xmlvalid.bat
```

symphorien replied:
then they should probably go in $out/libexec or something like that.

tbenst replied:
That will cause the software to break as there are hardcoded assumptions to the relative path

veprbl replied:
I agree with @symphorien . If we wanted to do things according to FHS then bf.sh would go to libexec/ and the jar file would probably go to share/java/. The xml file is probably just garbage.

veprbl commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,36 @@
+    cp ./* $out/
+    for script in $(find . -maxdepth 1 -perm -111 -type f); do
+      ln -s $out/$script $out/bin/$script
+    done
```
Probably should simply copy everything except *.bat files to $out/bin.

tbenst replied:
ah ok!

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -964,6 +964,8 @@ in
 
   bfr = callPackage ../tools/misc/bfr { };
 
+  bftools = callPackage ../applications/science/biology/bftools { };
```
In principle there is whole section for SCIENCE/BIOLOGY down there:
https://github.com/NixOS/nixpkgs/blob/6bc8a6621296d073974c4537933b53aa029edde5/pkgs/top-level/all-packages.nix#L21415

veprbl commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,36 @@
+        done
+      '';
+
+  buildInputs = [ unzip pkgs.makeWrapper ];
```
pkgs.makeWrapper should go to `nativeBuildInputs`

veprbl replied:
This is funny. I don't understand too much of this either.

In essence, things that you put into `nativeBuildInputs` will be some native utilities that are supposed to run during the build (so they are "native" to the architecture where build executes) and nativeBuildInputs is for dependencies (in case of cross compilation they will be the ones that were also compiled for the other architecture).

tbenst replied:
quick clarification, do you mean "buildInputs is for dependencies"?

veprbl replied:
@tbenst Yes, corrected, thanks.

tbenst committed:
bftools: init at 5.9.2

tbenst commented:
Thanks for the great feedback, @veprbl! I've pushed corrections for all resolved conversations

veprbl commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,34 @@
+  meta = with stdenv.lib; {
+    description = "A bundle of scripts for using Bio-Formats on the command line with bioformats_package.jar already included";
+    license = licenses.gpl3;
+    meta.platforms = platforms.all;
```
```suggestion
    platforms = platforms.all;
```

veprbl commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,34 @@
+{ stdenv, lib, makeWrapper, fetchurl, unzip, openjdk }:
```
You should use `openjre`, not `openjdk` unless bftools need to use java compiler.

veprbl commented:
Also you don't need unzip:
```diff
diff --git a/pkgs/applications/science/biology/bftools/default.nix b/pkgs/applications/science/biology/bftools/default.nix
index aa30cbfc5ca..c45a577fc60 100644
--- a/pkgs/applications/science/biology/bftools/default.nix
+++ b/pkgs/applications/science/biology/bftools/default.nix
@@ -1,12 +1,12 @@
-{ stdenv, lib, makeWrapper, fetchurl, unzip, openjdk }:
+{ stdenv, lib, makeWrapper, fetchzip, unzip, openjdk }:

 stdenv.mkDerivation rec {
   name = "bftools-${version}";
   version = "5.9.2";

-  src = fetchurl {
+  src = fetchzip {
     url = "http://downloads.openmicroscopy.org/bio-formats/${version}/artifacts/bftools.zip";
-    sha256 = "02x8p07z7js41kgiblkr9k8v9v4ka8rm7pg3f1vy9s5p6hmpfvnx";
+    sha256 = "08lmbg3kfxh17q6548il0i2h3f9a6ch8r0r067p14dajhzfpjyqj";
   };

   installPhase = ''
@@ -20,8 +20,6 @@ stdenv.mkDerivation rec {
     done
   '';

-  buildInputs = [ unzip ];
-
   nativeBuildInputs = [ makeWrapper ];

   meta = with stdenv.lib; {

```

veprbl committed:
bftools: init at 5.9.2

Co-Authored-By: tbenst <tbenst@gmail.com>

veprbl commented on pkgs/applications/science/biology/bftools/default.nix:
```diff
@@ -0,0 +1,34 @@
+
+  meta = with stdenv.lib; {
+    description = "A bundle of scripts for using Bio-Formats on the command line with bioformats_package.jar already included";
+    license = licenses.gpl3;
```
```suggestion
    license = licenses.gpl2;
```
https://github.com/openmicroscopy/bioformats/blob/1b39cedbc7744731f1c40cebfef3182ac88c11d2/LICENSE.txt
http://downloads.openmicroscopy.org/bio-formats/5.9.2/artifacts/bioformats-5.9.2.tar.xz

veprbl committed:
bftools: init at 5.9.2

Co-Authored-By: tbenst <tbenst@gmail.com>

tbenst committed:
bftools: init at 5.9.2

tbenst committed:
bftools: init at 5.9.2

Merge branch 'bioformattools' of github.com:/tbenst/nixpkgs into bioformattools

tbenst commented:
without unzip, I get `do not know how to unpack source archive /nix/store/qiq3mf2jdlafc35bqkw8npzpzwnj822z-bftools.zip`, so I kept it.

Made all other changes except @symphorien's idea, as I'm not sure how to change the directory structure without breaking the software.

veprbl commented:
@tbenst This is why my snippet switches `fetchurl` to `fetchzip`.

tbenst committed:
bftools: init at 5.9.2

tbenst commented:
@vberger doh! fixed

veprbl commented:
Switching the directory structure would require you to patch the shell scripts. Something like:
```bash
substituteInPlace "$out"/bin/bfconvert \
  --replace "\$BF_DIR" "$out"/libexec
```

tbenst commented:
thanks! I got stuck for now and can't spend more time on this for a few days:

```
  installPhase = ''
    mkdir -p $out/bin
    mkdir $out/libexec
    mkdir -p $out/share/java
    cp $(find . -maxdepth 1 -perm -111 -type f) $out/bin
    cp ./*.sh $out/libexec
    cp ./*.jar $out/share/java
    for file in $out/bin/*; do
      substituteInPlace $file --replace "\$BF_DIR" $out/libexec
      substituteInPlace $file --replace "\$BF_JAR_DIR" $out/share/java
    done
  '';
```

```
$ bfconvert 
Required JAR libraries not found. Please download:
  bioformats_package.jar
from:
  https://downloads.openmicroscopy.org/latest/bio-formats/artifacts/
and place in the same directory as the command line tools.
```

tbenst committed:
bftools: init at 5.9.2

veprbl commented:
@GrahamcOfBorg build bftools

veprbl committed:
bftools: fix runtime, refactor installPhase

veprbl commented:
@GrahamcOfBorg build bftools
